src/sinks/gcp/stackdriver_metrics.rs

```diff
@@ -92,162 +96,325 @@ fn default_endpoint() -> String {
+            })?
+            .connect_lazy();
+
+        let mut stopwatch = tokio::time::interval(Duration::from_secs(1));
+        let mut aggregate = HashMap::<String, MetricDataPoint>::with_capacity(GCP_MAX_BATCH_SIZE);
+        let mut time_tracker = StartTimeTracker::new();
+        let mut last_time = Instant::now();
+        let mut buffer = Vec::<MetricDataPoint>::new();
+        let mut client =
+            MetricServiceClient::<Channel>::with_interceptor(channel, interceptor(auth));
+
+        loop {
+            if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                while last_time.elapsed() < GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY {
+                    tokio::time::sleep(Duration::from_millis(500)).await;
+                }
+            } else {
+                while let Some(data_point) = buffer.pop() {
+                    aggregate
+                        .entry(format!("{}/{}", data_point.namespace, data_point.name))
+                        .and_modify(|cur| cur.merge(&data_point))
+                        .or_insert(data_point);
+
+                    if aggregate.len() == GCP_MAX_BATCH_SIZE {
+                        break;
+                    }
+                }
+            }
 
-    async fn build_request(
-        &self,
-        mut metrics: Self::Output,
-    ) -> crate::Result<hyper::Request<Bytes>> {
-        let metric = metrics.pop().expect("only one metric");
-        let (series, data, _metadata) = metric.into_parts();
-        let namespace = series
-            .name
-            .namespace
-            .unwrap_or_else(|| self.config.default_namespace.clone());
-        let metric_type = format!(
-            "custom.googleapis.com/{}/metrics/{}",
-            namespace, series.name.name
-        );
-
-        let end_time = data.time.timestamp.unwrap_or_else(chrono::Utc::now);
-
-        let (point_value, interval, metric_kind) = match &data.value {
-            MetricValue::Counter { value } => {
-                let interval = gcp::GcpInterval {
-                    start_time: Some(self.started),
-                    end_time,
+            if !aggregate.is_empty()
+                && last_time.elapsed() >= GCP_SMALLEST_CREATE_TIMESERIES_CALL_FREQUENCY
+            {
+                let start_time = time_tracker.start_time();
+                let time_series = aggregate
+                    .drain()
+                    .map(|t| gcp_timeseries(&self.config, start_time, t.1))
+                    .collect::<Vec<_>>();
+
+                let req = googleapis_proto::CreateTimeSeriesRequest {
+                    name: format!("projects/{}", self.config.project_id),
+                    time_series,
                 };
 
-                (*value, interval, gcp::GcpMetricKind::Cumulative)
+                let mut attempts = 1usize;
+
+                loop {
+                    if let Err(status) = client.create_time_series(Request::new(req.clone())).await
```

Comment:
The `clone` here is only really necessary so we have the request again should an error occur and we want to resend - which should happen very rarely. 

Could we change this so we extract the code that creates `req` into a separate function and then just call that each time we need it, thus avoiding the need to `clone` for the happy path?

Reply:
I don't see how it's possible considering that we `drain` the `aggregate` map we used to accumulate the data points. `clone()` is the only to achieve it.

All things consider, this is a pretty rare situation and I'm perfectly fine if we just skip the retry part entirely.

Reply:
Retries should probably be handled by the stream sink driver anyways.